### PR TITLE
PAL_VirtualUnwindOutOfProc for MacOS

### PR DESCRIPTION
**Pull request by mikem8361**
Add MacOS compact unwind section parsing (__unwind_info) and stepping: SearchCompactEncodingSection, SearchDwarfSection, GetProcInfo, StepWithCompactEncoding*.

Add back some (from a previous commit) of the dwarf unwind info parsing functions to use for the dwarf unwind info (__eh_frame section): ReadValue*, ReadULEB128, ReadSLEB128, ReadEncodedPointer, ParseCie, ExtractFde, ExtractProcInfoFromFde. These functions were used and fairly well tested in 3.x.

Build libunwind8 source on MacOS for those dwarf unwind info cases.

### CODE REVIEW

**Comment by mikem8361** [comment]
@janvorli, @sdmaclea as far as the changes I made to build libunwind8 for MacOS, I hope we can come up with a way to add them without requiring them to be pushed to the libunwind8 repo.  I didn't make any changes to the core files only changed CMakeLists.txt and adding some subdirectories (which could be moved out of the libunwind directory).

File: src/coreclr/src/debug/daccess/dacfn.cpp
Comment (sdmaclea) [inline src/coreclr/src/debug/daccess/dacfn.cpp:282]: This looks fishy.  What if `!FEATURE_DATATARGET4` and `SUCCEEDED(hr)`

File: src/coreclr/src/pal/src/exception/compact_unwind_encoding.h
Comment (sdmaclea) [inline src/coreclr/src/pal/src/exception/compact_unwind_encoding.h:3]: Is this correct?  Did we edit LLVM's file?  Do we need to add the .NET foundation header?

File: src/coreclr/src/pal/src/libunwind/src/CMakeLists.txt
Comment (sdmaclea) [inline src/coreclr/src/pal/src/libunwind/src/CMakeLists.txt:339]: I assume this is intentional and the cross DAC still compiles.

File: src/coreclr/src/pal/src/libunwind_mac/include/elf.h
Comment (sdmaclea) [inline src/coreclr/src/pal/src/libunwind_mac/include/elf.h:2]: Some of these files appear to be direct clones of the Windows platform abstraction files I added to make Windows libunwind compile.  I'm not sue how I feel about cloning them again.

I guess it really depends on whether we plan to upstream this.

File: src/coreclr/src/pal/src/libunwind_mac/src/missing-functions.c
Comment (sdmaclea) [inline src/coreclr/src/pal/src/libunwind_mac/src/missing-functions.c:3]: I wonder if this should use libunwind's license header.

**Review by sdmaclea — COMMENTED** [review]
It is not inherently obvious why we wouldn't want to upstream this.  Perhaps because MacOS libunwind support is always installed.  So there is no need for MacOS support in the libunwind project.

I have mixed feelings about the directory layout.  I would wait for @janvorli feedback on the directory style.

File: src/coreclr/src/debug/daccess/dacfn.cpp
Comment (mikem8361) [reply src/coreclr/src/debug/daccess/dacfn.cpp:282]: Then hr is E_NOINTERFACE and PAL_VirtualUnwindOutOfProc is called.

What SUCCEEDED(hr) are you looking at? 

```

    HRESULT hr = E_NOINTERFACE;           

#ifdef FEATURE_DATATARGET4
    ReleaseHolder<ICorDebugDataTarget4> dt;
    hr = g_dacImpl->m_pTarget->QueryInterface(IID_ICorDebugDataTarget4, (void **)&dt);
    if (SUCCEEDED(hr))
    {
        hr = dt->VirtualUnwind(threadId, sizeof(CONTEXT), (BYTE*)context);
    }
#endif

    if (hr == E_NOINTERFACE || hr == E_NOTIMPL)
    {
        hr = S_OK;

        SIZE_T baseAddress = DacGlobalBase();
        if (baseAddress == 0 || !PAL_VirtualUnwindOutOfProc(context, contextPointers, baseAddress, DacReadAllAdapter))
        {
            hr = E_FAIL;
        }
    }
```

File: src/coreclr/src/pal/src/exception/compact_unwind_encoding.h
Comment (mikem8361) [reply src/coreclr/src/pal/src/exception/compact_unwind_encoding.h:3]: I added the MIT license.  Not sure if I should have or not.

File: src/coreclr/src/pal/src/libunwind/src/CMakeLists.txt
Comment (mikem8361) [reply src/coreclr/src/pal/src/libunwind/src/CMakeLists.txt:339]: It does.  The reason is I added a copy of the two functions remote-unwind.cpp needed from seh-unwind.cpp to remote-unwind.cpp.

File: src/coreclr/src/pal/src/libunwind_mac/src/missing-functions.c
Comment (mikem8361) [reply src/coreclr/src/pal/src/libunwind_mac/src/missing-functions.c:3]: I'm not sure either.  I did copy some of those function from the libunwind source.

**Comment by mikem8361** [comment]
As far as push these changes to the libunwind repo: it would require a lot of work because most of the time the dwarf unwind isn't used so libunwind isn't used.  It is the compact/compressed unwind info that is used at lot more and requires the compact "step" code I wrote in remote-unwind.cpp. So most of the new code would have to be ported to the libunwind8 coding conventions, etc.  In other words, without the compact format parsing/stepping just building it on macOS  doesn't have any benefit.

File: src/coreclr/src/pal/src/libunwind_mac/include/elf.h
Comment (mikem8361) [reply src/coreclr/src/pal/src/libunwind_mac/include/elf.h:2]: The windows versions of these duplicate files are ifdef'ed _MSC_VER and I didn't want to change them.  We can fix that if we do actually ever upstream these changes.

File: src/coreclr/src/pal/src/libunwind_mac/include/endian.h
Comment (janvorli) [inline src/coreclr/src/pal/src/libunwind_mac/include/endian.h:1]: Do we need these fake files for Mac at all? On Windows, we didn't have some of the headers / symbols / functions. But on OSX, I would expect most if not all of those being present in the platform headers.

File: src/coreclr/src/pal/src/exception/compact_unwind_encoding.h
Comment (mikem8361) [reply src/coreclr/src/pal/src/exception/compact_unwind_encoding.h:3]: I'll remove the MIT header I added.

File: src/coreclr/src/pal/src/exception/remote-unwind.cpp
Comment (mikem8361) [reply src/coreclr/src/pal/src/exception/remote-unwind.cpp:167]: I added a comment.

File: src/coreclr/src/pal/src/libunwind_mac/src/missing-functions.c
Comment (mikem8361) [reply src/coreclr/src/pal/src/libunwind_mac/src/missing-functions.c:3]: I removed the MIT header and added the libunwind one.

File: src/coreclr/src/pal/src/CMakeLists.txt
Comment (mikem8361) [reply src/coreclr/src/pal/src/CMakeLists.txt:242]: I'll make these changes.

File: src/coreclr/src/pal/src/libunwind_mac/include/endian.h
Comment (mikem8361) [reply src/coreclr/src/pal/src/libunwind_mac/include/endian.h:1]: Some of them like elf.h doesn't exist on macOS. Others, in order to build libunwind without any changes are needed to fake it out.

**Review by sdmaclea — APPROVED** [review]

File: src/coreclr/src/pal/src/CMakeLists.txt
Comment (janvorli) [inline src/coreclr/src/pal/src/CMakeLists.txt:251]: Just a final nit - can you please put one directory per line here instead of the semicolon separated list and use `CMAKE_CURRENT_SOURCE_DIR` instead of `${CLR_DIR}/src/pal/src` as follows?
```suggestion
    ${CMAKE_CURRENT_SOURCE_DIR}/libunwind/include
    ${CMAKE_CURRENT_SOURCE_DIR}/libunwind/include/tdep
    ${CMAKE_CURRENT_BINARY_DIR}
    ${CMAKE_CURRENT_BINARY_DIR}/libunwind/include
    ${CMAKE_CURRENT_BINARY_DIR}/libunwind/include/tdep```

**Review by janvorli — APPROVED** [review]
LGTM modulo the nit.

File: src/coreclr/src/pal/src/CMakeLists.txt
Comment (mikem8361) [reply src/coreclr/src/pal/src/CMakeLists.txt]: Done.

**Pull request merged**
